Mathlib/Topology/Gluing.lean

```diff
@@ -0,0 +1,542 @@
+    ext x
+    conv_rhs => rw [← Set.preimage_image_eq U (D.ι_injective _)]
+    generalize 𝖣.ι i '' U = U'
+    simp
+  rw [← this, Set.image_preimage_eq_inter_range]
+  symm
+  apply Set.inter_eq_self_of_subset_left
+  rw [← D.preimage_range i j]
+  exact Set.preimage_mono (Set.image_subset_range _ _)
+set_option linter.uppercaseLean3 false in
+#align Top.glue_data.preimage_image_eq_image TopCat.GlueData.preimage_image_eq_image
+
+theorem preimage_image_eq_image' (i j : D.J) (U : Set (𝖣.U i)) :
+    𝖣.ι j ⁻¹' (𝖣.ι i '' U) = (D.t i j ≫ D.f _ _) '' (D.f _ _ ⁻¹' U) := by
+  convert D.preimage_image_eq_image i j U using 1
+  rw [coe_comp, coe_comp]
+  -- porting note: `show` was not needed, since `rw [← Set.image_image]` worked.
+  show (fun x => ((forget TopCat).map _ ((forget TopCat).map _ x))) '' _ = _
+  rw [← Set.image_image]
+  -- porting note: `congr 1` was here, instead of `congr_arg`, however, it did nothing.
+  refine congr_arg ?_ ?_
+  rw [← Set.eq_preimage_iff_image_eq, Set.preimage_preimage]
+  change _ = (D.t i j ≫ D.t j i ≫ _) ⁻¹' _
+  rw [𝖣.t_inv_assoc]
+  rw [← isIso_iff_bijective]
+  apply (forget TopCat).map_isIso
+set_option linter.uppercaseLean3 false in
+#align Top.glue_data.preimage_image_eq_image' TopCat.GlueData.preimage_image_eq_image'
+
+-- porting note: the goal was simply `IsOpen (𝖣.ι i '' U)`.
+-- I had to manually add the explicit type ascription.
+theorem open_image_open (i : D.J) (U : Opens (𝖣.U i)) : IsOpen (𝖣.ι i '' (U : Set (D.U i))) := by
+  rw [isOpen_iff]
+  intro j
+  rw [preimage_image_eq_image]
+  apply (D.f_open _ _).isOpenMap
+  apply (D.t j i ≫ D.f i j).continuous_toFun.isOpen_preimage
+  exact U.isOpen
+set_option linter.uppercaseLean3 false in
+#align Top.glue_data.open_image_open TopCat.GlueData.open_image_open
+
+theorem ι_openEmbedding (i : D.J) : OpenEmbedding (𝖣.ι i) :=
+  openEmbedding_of_continuous_injective_open (𝖣.ι i).continuous_toFun (D.ι_injective i) fun U h =>
+    D.open_image_open i ⟨U, h⟩
+set_option linter.uppercaseLean3 false in
+#align Top.glue_data.ι_open_embedding TopCat.GlueData.ι_openEmbedding
+
+/-- A family of gluing data consists of
+1. An index type `J`
+2. A bundled topological space `U i` for each `i : J`.
+3. An open set `V i j ⊆ U i` for each `i j : J`.
+4. A transition map `t i j : V i j ⟶ V j i` for each `i j : ι`.
+such that
+6. `V i i = U i`.
+7. `t i i` is the identity.
+8. For each `x ∈ V i j ∩ V i k`, `t i j x ∈ V j k`.
+9. `t j k (t i j x) = t i k x`.
+
+We can then glue the topological spaces `U i` together by identifying `V i j` with `V j i`.
+-/
+-- Porting note: removed `@[nolint has_nonempty_instance]`
+structure MkCore where
+  {J : Type u}
+  U : J → TopCat.{u}
+  V : ∀ i, J → Opens (U i)
+  t : ∀ i j, (Opens.toTopCat _).obj (V i j) ⟶ (Opens.toTopCat _).obj (V j i)
+  V_id : ∀ i, V i i = ⊤
+  t_id : ∀ i, ⇑(t i i) = id
+  -- porting note: was vv `@coe (V j i) (U j) _ (t i j x) ∈ V j k`
```

Comment:
What is the `vv` here? I'm still confused.

Reply:
Oh, is it meant to be an arrowhead pointing at something on the next line? That is way too confusing for me.

Reply:
(That would explain why previously this line had weird indenting?)

Reply:
```suggestion
```